overlord/snapstate/backend/link.go

```diff
@@ -143,20 +143,57 @@ func (b Backend) StopServices(apps []*snap.AppInfo, reason snap.ServiceStopReaso
 	return wrappers.StopServices(apps, reason, meter, tm)
 }
 
-func generateWrappers(s *snap.Info) error {
+func generateWrappers(s *snap.Info, disabledSvcs []string) error {
+	var err error
 	// add the CLI apps from the snap.yaml
-	if err := wrappers.AddSnapBinaries(s); err != nil {
+	if err = wrappers.AddSnapBinaries(s); err != nil {
 		return err
 	}
-	// add the daemons from the snap.yaml
-	if err := wrappers.AddSnapServices(s, progress.Null); err != nil {
+	// cleanup if we fail
+	defer func() {
+		if err == nil {
+			return
+		}
 		wrappers.RemoveSnapBinaries(s)
+	}()
+
+	// add the daemons from the snap.yaml
+	if err = wrappers.AddSnapServices(s, progress.Null); err != nil {
```

Comment:
as I tried to explain, I think this one should take disabledApps (possibly empty)  as well instead of enabling here and disabling shortly after with a different helper